sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -287,76 +289,145 @@ private SpecializedBlobClientBuilder prepareBuilder() {
-                .concatWith(Flux.defer(pool::flush))
-                .flatMapSequential(buffer -> {
-                    // Report progress as necessary.
-                    Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(Flux.just(buffer),
-                        finalParallelTransferOptions.getProgressReceiver(), progressLock, totalProgress);
-
-                    final String blockId = Base64.getEncoder().encodeToString(
-                        UUID.randomUUID().toString().getBytes(UTF_8));
-
-                    return getBlockBlobAsyncClient().stageBlockWithResponse(blockId, progressData, buffer.remaining(),
-                        null, accessConditionsFinal.getLeaseId())
-                        // We only care about the stageBlock insofar as it was successful,
-                        // but we need to collect the ids.
-                        .map(x -> blockId)
-                        .doFinally(x -> pool.returnBuffer(buffer))
-                        .flux();
-                }) // TODO: parallelism?
-                .collect(Collectors.toList())
-                .flatMap(ids ->
-                    getBlockBlobAsyncClient()
-                        .commitBlockListWithResponse(ids, headers, metadata, tier, accessConditions));
-        } catch (RuntimeException ex) {
-            return monoError(logger, ex);
-        }
+        return chunkedSource.concatMap(pool::write)
+            .concatWith(Flux.defer(pool::flush))
+            .flatMapSequential(buffer -> {
+                // Report progress as necessary.
+                Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(Flux.just(buffer),
+                    parallelTransferOptions.getProgressReceiver(), progressLock, totalProgress);
+
+                final String blockId = Base64.getEncoder().encodeToString(
+                    UUID.randomUUID().toString().getBytes(UTF_8));
+
+                return blockBlobAsyncClient.stageBlockWithResponse(blockId, progressData, buffer.remaining(),
+                    null, accessConditions.getLeaseId())
+                    // We only care about the stageBlock insofar as it was successful,
+                    // but we need to collect the ids.
+                    .map(x -> blockId)
+                    .doFinally(x -> pool.returnBuffer(buffer))
+                    .flux();
+            }) // TODO: parallelism?
+            .collect(Collectors.toList())
+            .flatMap(ids ->
+                blockBlobAsyncClient.commitBlockListWithResponse(ids, headers, metadata, tier, accessConditions));
+    }
+
+    private Mono<Response<BlockBlobItem>> determineUploadFullOrChunked(final Flux<ByteBuffer> data,
+        final Function<Flux<ByteBuffer>, Mono<Response<BlockBlobItem>>> uploadInChunks,
+        final BiFunction<Flux<ByteBuffer>, Long, Mono<Response<BlockBlobItem>>> uploadFullBlob) {
+        final long chunkedUploadRequirement = 4 * Constants.MB;
```

Comment:
Are we going to make 4 MB the default size to decide whether to chunk or not? If so, make this a static constant. I guess this way, the caching of BB is controlled by us and the overhead is limited to a max of 8 MB.
The other option is to use the block size in `ParallelTransferOptions` which can be 1 GB and we'll have to cache it resulting in a noticeable increase in memory usage.

Reply:
For now this will be made into a static constant.

Reply:
Even if we stick with this 4 megabyte value, what happens if block size in parallel transfer options is higher than this? This looks as if the following steps are possible

1. Try to upload 1023MB blob with 1GB block size
2. See that blob is larger than 4MB and then move to chunked
3. Upload the entire blob in a put block call and then commit block list

I don't know if it's the right decision to do all this work for smart branching only to wind up in this scenario. If we allow a block size as high as 1GB then we should support it throughout our decision making.

Reply:
I think @srnagar slightly misspoke, the maximum allowed for a stage block call is 100MB.

It is a good point about using Put Blob up to the limit of a passed buffer size. I would have a slight concern if someone accidentally passed 10KB instead of 10MB and other order of magnitude mistakes, possibly doing a Math.max(4MB, bufferSize).

Reply:
Right about the 100MB. My mind just went there because of the 1GB remark.

Anyway, I do agree with Rick in that we should be listening to the customer's configuration settings if they are provided, and using this constant here seems like we're ignoring those values for our decision making.

Reply:
I'm good with making the change to allow the `blockSize` configuration to determine which code path is taken, will make sure to update the Javadocs for this method to indicate that its value will determine whether Put Blob or Stage Blocks is used.

Reply:
Updated to use `blockSize` to determine the Put Blob threshold.

Reply:
Wait sorry for the churn. I think we want this to have the same putBlob vs. putBlock behavior as uploadFromFile for consistency between APIs, and uploadFromFile should have a separate value in the parallelTransferOptions, otherwise they can't make use of the 256mb limit on putBlob. My suggestion earlier of configuration was meant as a future work item to be alongside when we add this for uploadFromFile. I propose we always use the default for the blocksize now, and we'll add configuration in as a feature for both apis later.

Reply:
Ah, I'll revert to the last commit.

Reply:
Issue to add this option to `ParallelTransferOptions` https://github.com/Azure/azure-sdk-for-java/issues/6097